src/models/gamma_poisson_model.py

```diff
@@ -0,0 +1,402 @@
+
+    def _kreach(self, k, I, Imax, alpha, beta):
+        """Probability that a random user receives k impressions.
+
+        Computes the function:
+            sum_{n=k}^{infinity} g(k, n | alpha, beta, I, I_max).
+        where g(k, n | alpha, beta, I, I_max) is the probability that
+        a randomly chosen user will have an available inventory of n
+        impressions, of which k are shown.
+
+        Args:
+          k:  np.array specifying number of impressions that were seen by the user.
+          I:  Total number of impressions shown to all users.
+          Imax: Total size of impression inventory.
+          alpha: Parameter of Gamma-Poisson distribution.
+          beta: Parameter of Gamma-Poisson distribution.
+        Returns:
+          For each k, probability that a randomly chosen user will have an inventory
+          of n impressions, of which k are shown.
+        """
+        return np.array(
+            [
+                np.sum(
+                    self._knreach(
+                        kv,
+                        np.arange(kv, MAXIMUM_COMPUTATIONAL_FREQUENCY + 1),
+                        I,
+                        Imax,
+                        alpha,
+                        beta,
+                    )
+                )
+                for kv in k
+            ]
+        )
+
+    def _expected_impressions(self, N, alpha, beta):
+        """Estimates the expected size of impression inventory for N users.
+
+        Makes use of the identity that the expected value of a negative binomial
+        distribution with parameters p and r is pr/(1-p).
+
+        Args:
+          N:      total number of people that can potentially be reached.
+          alpha:  shape parameter of Gamma distribution.
+          beta:   rate parameter of Gamma distribution.
+        Returns:
+          Expected size of impression inventory for a population of N users.
+        """
+        return N * alpha / beta
+
+    def _expected_histogram(self, I, Imax, N, alpha, beta, max_freq=10):
+        """Computes an estimated histogram.
+
+        Args:
+          I:      number of impressions that were purchased.
+          Imax:   total size of impression inventory.
+          N:      total number of people that can potentially be reached.
+          alpha:  shape parameter of Gamma distribution.
+          beta:   rate parameter of Gamma distribution.
+          max_freq: maximum frequency to report in the output histogram.
+        Returns:
+          An np.array h where h[i] is the expected number of users who will see i+1
+          impressions.
+        """
+        return N * self._kreach(np.arange(1, max_freq + 1), I, Imax, alpha, beta)
+
+    def _impression_count(self, h):
+        """Number of impressions recorded in the histogram h."""
+        return np.sum([h[i] * (i + 1) for i in range(len(h))])
+
+    def _feasible_point(self, h):
+        """Returns values of alpha, beta, Imax that could feasibly produce h."""
```

Comment:
It would be great if there is some comment explaining the intuition behind the setting of ```alpha```, ```beta```, and ```I```.